src/core/allow-bind.c

```diff
@@ -0,0 +1,123 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "allow-bind.h"
+#include "unit.h"
+
+#if BUILD_BPF
+/* libbpf, clang and llc compile time dependencies are satisfied */
+#include "bpf-object.h"
+#include "bpf/allow_bind/allow-bind-hexdump.h"
+#include "cgroup-bpf.h"
+#include "set.h"
+
+static int prepare_bpf_object(Unit* u, const Set *ports, struct bpf_object **object) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        const char *map_name = "allow_ports";
+        int map_fd, r;
+        uint8_t dummy;
+        void *port;
+
+        assert(ports);
+        assert(object);
+
+        r = bpf_object_new(allow_bind_hexdump_buffer, sizeof(allow_bind_hexdump_buffer), &obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to create BPF object from hexdump buffer: %m");
+
+        if (set_size(ports) > 0) {
+                r = bpf_object_resize_map(obj, map_name, set_size(ports));
+                if (r < 0)
+                        return log_unit_error_errno(u, r, "Failed to resize BPF map '%s': %m", map_name);
+        }
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to load BPF object: %m");
+
+        map_fd = bpf_object_get_map_fd(obj, map_name);
+        if (map_fd < 0)
+                return log_unit_error_errno(u, r, "Failed to get BPF map '%s' fd: %m", map_name);
+
+        SET_FOREACH(port, ports)
+                if (bpf_map_update_elem(map_fd, &port, &dummy, BPF_ANY))
+                        return log_unit_error_errno(u, r, "Failed to update BPF map '%s", map_name);
+
+        *object = TAKE_PTR(obj);
+        return 0;
+}
+
+int allow_bind_supported(void) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        int r;
+
+        r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
+        if (r < 0)
+                return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
+
+        if (r == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOTSUP),
```

Comment:
We use EOPNOTSUPP, not ENOTSUP, for some reason that I was explained once and that I promptly forgot :-)

Reply:
also maybe return 0 here? to match the case where it's not built with bpf? they are all missing requirements, right?

Reply:
@bluca 

`supported` helper now returns `0` if requirements are not met.